cmd/blob/acceptance_test.go

```diff
@@ -0,0 +1,110 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	as "github.com/miquido/conduit-connector-azure-storage"
+	asSource "github.com/miquido/conduit-connector-azure-storage/source"
+	helper "github.com/miquido/conduit-connector-azure-storage/test"
+	"go.uber.org/goleak"
+)
+
+type CustomConfigurableAcceptanceTestDriver struct {
+	sdk.ConfigurableAcceptanceTestDriver
+
+	containerClient *azblob.ContainerClient
+}
+
+func (d *CustomConfigurableAcceptanceTestDriver) GenerateRecord(t *testing.T) sdk.Record {
+	record := d.ConfigurableAcceptanceTestDriver.GenerateRecord(t)
+
+	// Override Key
+	record.Key = sdk.RawData(fmt.Sprintf("file-%d.txt", time.Now().UnixMicro()))
+
+	// Override CreatedAt
+	// azurite does not support milliseconds precision
+	record.CreatedAt = record.CreatedAt.Truncate(time.Second)
+
+	return record
+}
+
+func (d *CustomConfigurableAcceptanceTestDriver) WriteToSource(t *testing.T, records []sdk.Record) (results []sdk.Record) {
+	for _, record := range records {
+		_ = helper.CreateBlob(
+			d.containerClient,
+			string(record.Key.Bytes()),
+			"text/plain",
+			string(record.Payload.Bytes()),
+		)
+	}
+
+	// No destination connector, return wanted records
+	return records
+}
+
+func TestAcceptance(t *testing.T) {
+	sourceConfig := map[string]string{
+		asSource.ConfigKeyConnectionString: helper.GetConnectionString(),
+		asSource.ConfigKeyContainerName:    "acceptance-tests",
+	}
+
+	testDriver := CustomConfigurableAcceptanceTestDriver{
+		ConfigurableAcceptanceTestDriver: sdk.ConfigurableAcceptanceTestDriver{
+			Config: sdk.ConfigurableAcceptanceTestDriverConfig{
+				Connector: sdk.Connector{
+					NewSpecification: as.Specification,
+					NewSource:        asSource.NewSource,
+					NewDestination:   nil,
+				},
+
+				SourceConfig:     sourceConfig,
+				GenerateDataType: sdk.GenerateRawData,
+
+				GoleakOptions: []goleak.Option{
+					// Routines created by Azure client
+					goleak.IgnoreTopFunction("internal/poll.runtime_pollWait"),
+					goleak.IgnoreTopFunction("net/http.(*persistConn).writeLoop"),
+				},
+
+				Skip: []string{
+					// When restarting snapshot position, bucket is read from the beginning
+					"TestAcceptance/TestSource_Open_ResumeAtPositionSnapshot",
+
+					// Cannot be tested. driver.WriteToSource() creates all records at once while CDC position is
+					// created after the Snapshot iterator finishes iterating. So from the CDC iterator point of view,
+					// there are no new records available.
+					"TestAcceptance/TestSource_Open_ResumeAtPositionCDC",
```

Comment:
Can you clarify this a little bit further please? Also, what's the actual error you're getting?

 [TestSource_Open_ResumeAtPositionCDC](https://github.com/ConduitIO/conduit-connector-sdk/blob/main/acceptance_testing.go#L703-L724) is doing the following:
* opens a source, which has no records in it
* reads all the records
* acknowledges half of those
* restarts the source

Regardless of the internals of a connector, this use case is something which should pass.

Reply:
@lovromazgon also pointed out that [these 4 lines](https://github.com/ConduitIO/conduit-connector-sdk/blob/a2f9ef8f08abc37587cdee7829935809b4af8d27/acceptance_testing.go#L706-L710) should give the source enough time to switch from doing a snapshot (without data) and switch to CDC. The timeout was lower in one of the previous versions of the SDK, so that might have been an issue for you. What he suggested is to upgrade to the latest SDK.

Reply:
It is not the matter of time required for connector to switch from the Snapshot iterator to CDC. To make this test working, you would need to change the TestSource_Open_ResumeAtPositionCDC test logic:
<img width="765" alt="obraz" src="https://user-images.githubusercontent.com/5360041/180411405-aa570c6d-2908-4e9c-bc70-4051c5f29bff.png">

`want` creates all files in the source and snapshot iterator sees them all. CDC iterator does not see any changes.

Reply:
> want creates all files in the source and snapshot iterator sees them all. CDC iterator does not see any changes.

I think here we have the answer. The expectation in this test is that there's an empty snapshot, after which the connector should switch to CDC. i.e. we have:
1. open source, which has no data in it
2. start reading -> no data, so no snapshot, and we start listening for changes.

Reply:
The important thing to note in that test is that it first tries to read something from the source and expects to get back a `context.DeadlineExceeded` or `sdk.ErrBackoffRetry` (see [these lines](https://github.com/ConduitIO/conduit-connector-sdk/blob/a2f9ef8f08abc37587cdee7829935809b4af8d27/acceptance_testing.go#L706-L710)). This means that when the connector first tries to do a snapshot it should discover that there is no data, meaning that the initial snapshot is empty and it should switch to CDC. Any new record that is created after that is considered a change of the existing state and not part of the existing state (existing state is empty).

We need a read timeout because we can't wait forever. The `Read` method is expected to block until it has a record to return, and there should be no records, so it would block forever. You can adjust the timeout if 5 seconds (default) is not enough to determine that there is no initial records to put into the snapshot.

If you don't think that is correct, I'd be interested to hear what you think should be the criteria for switching from "snapshot" to "CDC"?

Reply:
@hariso, @lovromazgon,
The problem here is with Snapshot → CDC iterators switch logic. I've just tested it for S3 connector using the same acceptance test code and I faced the same issue since I did the switch logic in the same manner.

The test logic seems to be right, but connectors have unusual behaviour:
`TestSource_Open_ResumeAtPositionCDC`:
703. `source := a.openSource` - create S3/Azure connector with Snapshot iterator as default
709. `_, err := source.Read` with `err` expected to be `ErrBackoffRetry` - iterator is still Snapshot, it will change on next iteration
	a. https://github.com/ConduitIO/conduit-connector-s3/blob/main/source/iterator/combined_iterator.go#L73-L82
	b. https://github.com/miquido/conduit-connector-azure-storage/blob/main/source/iterator/combined.go#L78-L87
715. `want := a.driver.WriteToSource(t, a.generateRecords(t, 10))` - 10 files are stored on S3/Azure bucket
718. `_, err = a.readMany(ctx, t, source, len(want))` - all generated files are read
And now, if You add another `_, err = a.readMany(ctx, t, source, len(want))` here, it will give `func (c *CombinedIterator) HasNext` a time for change of underlying iterator to CDC. Instead:
740. `gotRemaining, err := a.readMany(ctx, t, source, len(want)-len(got))` runs iterator change, `HasNext` returns `false`, `func (s *Source) Read` returns `sdk.ErrBackoffRetry` and:
741. `is.NoErr(err)` fails

Reply:
Our S3 connector doesn't pass acceptance tests yet (we have an open issue for that https://github.com/ConduitIO/conduit-connector-s3/issues/62). If the same behavior is present in the S3 connector it is unexpected and will be adjusted.

I glanced at the code and seems like we are missing a check in `CombinedIterator.Next` that would switch to CDC if `SnapshotIterator` returns `ErrSnapshotIteratorIsStopped` ([here](https://github.com/miquido/conduit-connector-azure-storage/blob/1dc53a90881cd32092fbd3b92ea34e7b1146eba2/source/iterator/combined.go#L100)). I didn't spend too much on this so I might be wrong.

Reply:
The check for `ErrSnapshotIteratorIsStopped` is already present in Read method:
https://github.com/miquido/conduit-connector-azure-storage/blob/1dc53a90881cd32092fbd3b92ea34e7b1146eba2/source/source.go#L98-L102

Then converted to `ErrBackoffRetry` and retried. Ideally would be to update the S3 and this connector logic so that it changes to CDC iterator earlier. I don't think it is in the scope of this PR, though.

Reply:
So you are saying the scope of this PR is to run acceptance tests and ignore the tests that fail? I would expect the scope to include fixing the issues found by the acceptance tests (unless they are unreasonably big issues). But let's not argue about technicalities, if you want to fix it in a follow-up PR that's fine by me.

Reply:
I'm not sure I'm able to understand the connection between this tests failure and the fact that Azure nor S3 support true cursors/scrolls/CDC?

If we take a step back, this test is about using CDC after we are done with the snapshot (be it empty or not). The Azure Storage connector has a CDC mode, so we expect the connector to switch to that mode.

Reply:
We discussed this at a standup. It looks like the best path forward is to accept the acceptance tests as they are, and work on improving the snapshot/CDC behavior separately. The issue is [here](https://github.com/miquido/conduit-connector-azure-storage/issues/12).

Thanks @lovromazgon @donatorsky for sharing your thoughts!